README.md
MPU6050-WebPlot
===============

plotting the signal from MPU6050

Latest version: v0.5.5

Usage
-----

**For back-end:**

```
$ ./service_start.py [serial port] [baud rate] [polling timeout]
```

The data format received from deivce is 6 integers separated by comma in one line. For example, `123,456,789,-123,-456,-789`.

**For front-end:**

Connect to [http://localhost:8888/](http://localhost:8888) in web browser.

Preview
-------

![web plot](https://github.com/KuoE0/MPU6050-WebPlot/raw/master/preview.png)

Environment
-----------

- OS X 10.9
- Python 2.7.6

Requirement
-----------

**Python**

- pyserial 2.7
- tornado 3.1.1

```
$ pip install pyserial
$ pip install tornado
```

**JavaScript**

- NodeJS
- npm
- bower

```
$ npm install -g bower
$ bower install
```






CHANGELOG
Changelog
----------

**v0.5.5**

- add signal panel to show signal information
- add toggle feature for signal display control

**v0.5.2**

- add parameter to change the timeout of polling  
- fix the error result of moving average filter
- add dynamic window size control of moving average filter
- redesign the UI
	- add two control panel
	- reduce the height of plotting graph

**v0.5**

- add button to apply moving average filter

**v0.4**

- add button to start/pause/continue signal capture
- add button to clear all signal

**v0.3.2**

- add homepage handler in service

**v0.3**

- use bower to manage JavaScript packages
- rearrange the path structure
- WebSocket communication supported


**v0.2**

- add signal value tracking feature

**v0.1.1**

- initial UI design

**v0.1**

- receive 6 signals from MPU6050
	- X-axis acceleration
	- Y-axis acceleration
	- Z-axis acceleration
	- Roll angle velocity
	- Pitch angle velocity
	- Yaw angle velocity


signal_receive.py
#! /usr/bin/env python
# -*- coding: utf-8 -*-
# vim:fenc=utf-8
#
# Copyright © 2013 KuoE0 <kuoe0.tw@gmail.com>
#
# Distributed under terms of the MIT license.

"""

"""
import tornado.httpserver
import tornado.ioloop
import tornado.web
import tornado.websocket
import serial
import signal
import sys
import json

tornado_port = 8888

# create serial object
serial_port = sys.argv[1]
serial_baudrate = int(sys.argv[2])
ser = serial.Serial(serial_port, serial_baudrate, timeout=1)

# global variable
client = list() # list of websocket client
number_of_signal = 200
serial_pending = list()
signal_set = [[0] * 6] * number_of_signal
signal_type = ['x-acc', 'y-acc', 'z-acc', 'x-gyro', 'y-gyro', 'z-gyro']

# SIGINT handler to close serial connection
def handler_SIGINT(signum, frame):
    global ser
    print "Signal {0} happened!".format(signum)
    print "Serial connection closed..."
    ser.close()
    sys.exit()

signal.signal(signal.SIGINT, handler_SIGINT)

# receive signal with a non-blocking way
def recieve_signal():

    data = ""
    try:
        if ser.inWaiting() != 0:
            data = ser.readline()
            print data
    except Exception as e:
        error_msg = "Error reading from {0}".format(serial_port)
        template = "An exception of type {0} occured. Arguments:\n{1!r}"
        message = template.format(type(e).__name__, e.args)

        print error_msg, message

    return data

# parse out the signal value
def parse_pending(signal_string):

    if len(signal_string) == 0:
        return

    global signal_set

    try:
    # split by ',' and get first element
        values = [int(x) for x in signal_string.split(',')]
    except:
        values = None

    # push signal into list
    if values and len(values) == 6:
        signal_set.append(values)

# push signal data to client
def signal_tx():
    parse_pending(recieve_signal())
    ret_signal = signal_set[:min(number_of_signal, len(signal_set))]
    if len(signal_set):
        signal_set.pop(0)

    # fill the signal
    if len(ret_signal) < number_of_signal:
        ret_signal.extend([[0] * 6] * (number_of_signal - len(ret_signal)))
    # transpose the list
    ret_signal = zip(*ret_signal)

    ret = list()
    for i in xrange(6):
        ret.append({ 'data': [p for p in enumerate(ret_signal[i])], 'label': signal_type[i] })
    ret = json.dumps({ 'signal': ret })
    
    for cl in client:
        cl.write_message(ret)

# tornado websocket handler
class socket_handler(tornado.websocket.WebSocketHandler):
    def open(self):
        client.append(self)
    def on_close(self):
        client.remove(self)

application = tornado.web.Application([
    (r"/ws", socket_handler),
    ])

if __name__ == "__main__":
    #tell tornado to run signal_tx every 1 ms
    serial_loop = tornado.ioloop.PeriodicCallback(signal_tx, 1)
    serial_loop.start()

    application.listen(tornado_port)
    print "Starting server on port number {0}...".format(tornado_port)
    print "Open at http://localhost:{0}/".format(tornado_port)

    try:
        tornado.ioloop.IOLoop.instance().start()
    except KeyboardInterrupt:
        print 'Server closed...'


static/js/main.js

$(function() {

	var number_of_signal = 200;
	var plot = null;
	var ws = undefined;
	var signal_info = {};

	function insert_signal(name) {
		var panel = $('#signal-panel .ui.grid');
		var id = signal_info[name].label, color = signal_info[name].color;
		panel.append('<div id="signal-panel-' + id + '" class="row"><div class="five wide column"><div id="signal-toggle-' + id + '" class="ui toggle checkbox"><input type="checkbox" checked /><label><i style="color: ' + color + '" class="sign icon"></i>' + id + '</label></div></div><div class="three wide column"><input class="signal-value" readonly/></div></div>');
	}


	function init_draw(data) {

		plot = $.plot('#signal-plot', data, {
			series: { 
				lines: { show: true, lineWidth: 1.5 },
				shadowSize: 0 },
			crosshair: { mode: "x" },
			grid: { hoverable: true, autoHighlight: false },
			xaxis: { show: false },
			yaxis: { min: -36000, max: 36000 },
			legend: { show: false}
		});

		data = plot.getData();

		for (var i = 0; i < data.length; ++i) {
			signal_info[data[i].label] = {
				'idx': i,
				'label': data[i].label,
				'color': data[i].color,
				'show': true
			}
			insert_signal(data[i].label);
		}

		$("[id^='signal-toggle-']").click(function () {
			var data = plot.getData();
			var label = $(this).attr('id').replace(/signal-toggle-/, '');
			var idx = signal_info[label].idx;
			data[idx].lines.show = !data[idx].lines.show;
			signal_info[label].show = !signal_info[label].show;
			plot.setData(data);
			plot.draw();
		});

		$('.ui.checkbox').checkbox();

	}

    // Websocket connection
    ws = new WebSocket("ws://localhost:8888/ws");

    ws.onmessage = function (evt) {

		var data = JSON.parse(evt.data);
		data = data.signal;

		if (plot == null) {
			init_draw(data);
		}
		else {

			for (var i = 0; i < data.length; ++i) {
				var label = data[i].label;
				data[i]['lines'] = {};
				data[i].lines['show'] = signal_info[label].show;
				data[i]['color'] = signal_info[label].color;
			}
			// set data for plot
			plot.setData(data);
			// redraw graph
			plot.draw();
		}
		// uplate value of signal
		update_signal_value();
	};


	var latest_pos = null;
	var update_signal_value_timeout = null;

	function update_signal_value() {

		if (latest_pos == null) {
			return;
		}

		update_signal_value_timeout = null;

		var pos = latest_pos;
		var axes = plot.getAxes();

		if (pos.x < axes.xaxis.min || pos.x > axes.xaxis.max ||
			pos.y < axes.yaxis.min || pos.y > axes.yaxis.max) {
			return;
		}

		var dataset = plot.getData();

		for (var i = 0; i < dataset.length; ++i) {
			var label = dataset[i].label;
			var series = dataset[i];

			var idx = Math.floor(pos.x);

			var p1 = series.data[idx], p2 = series.data[idx + 1];
			var ret = null;

			ret = (p1 == null) ? p2[1] : p1[1];

			$('#signal-panel-' + label + ' .signal-value').val(ret.toFixed(0));
		}
	};

	$('#signal-plot').bind("plothover", function (event, pos, iten) {
		latest_pos = pos;
		if (!update_signal_value_timeout) {
			update_signal_value_timeout = setTimeout(update_signal_value, 50);
		}
	});

	$('#clear-btn').click(function () {
		ws.send("clear");
	});

	$('#status-btn').click(function () {
		// close connection
		if ($(this).hasClass('active')) {
			$(this).removeClass('active');
			$(this).empty();
			$(this).append("<i class='play icon'></i>");
			ws.send("pause");
		}
		// open connection
		else {
			$(this).addClass('active');
			$(this).empty();
			$(this).append("<i class='pause icon'></i>");
			ws.send("play");
		}
	});

	$('#moving-average-filter').click(function () {
		if ($(this).hasClass('active')) {
			$(this).removeClass('active');
            $('#MA-slider').toggle('slow');
            $('#MA-slider').attr('disabled', true);
			$('#MA-value').text('');
			ws.send("MAF 0");
		}
		else {
			var value = $('#MA-slider').val();
			$(this).addClass('active');
			$('#MA-value').text('(' + value + ')');
            $('#MA-slider').toggle('fast');
            $('#MA-slider').attr('disabled', false);
		}
	});

	$('#MA-slider').change(function () {
		var value = $(this).val();
		$('#MA-value').text('(' + value + ')');
		$('#MA-slider').val(value);
		ws.send("MAF " + value);
	});

    $('.ui.checkbox').checkbox();


});


service_start.py
#! /usr/bin/env python
# -*- coding: utf-8 -*-
# vim:fenc=utf-8
#
# Copyright © 2013 KuoE0 <kuoe0.tw@gmail.com>
#
# Distributed under terms of the MIT license.

"""

"""
import tornado.httpserver
import tornado.ioloop
import tornado.web
import tornado.websocket
import numpy as np
import serial
import signal
import sys
import json
import os

tornado_port = 8888

# create serial object
serial_port = sys.argv[1]
serial_baudrate = int(sys.argv[2])
callback_timeout = int(sys.argv[3])
ser = serial.Serial(serial_port, serial_baudrate, timeout=1)

# global variable
client = list() # list of websocket client
number_of_signal = 200
serial_pending = list()
tx_status = False
size_window_MA = 0
signal_set = [[0] * 6] * number_of_signal
last_signal_set = [[0] * 6] * number_of_signal
signal_type = ['x-acc', 'y-acc', 'z-acc', 'x-gyro', 'y-gyro', 'z-gyro']

# SIGINT handler to close serial connection
def handler_SIGINT(signum, frame):
    global ser
    print "Signal {0} happened!".format(signum)
    print "Serial connection closed..."
    ser.close()
    sys.exit()

signal.signal(signal.SIGINT, handler_SIGINT)

# receive signal with a non-blocking way
def recieve_signal():

    data = ""
    try:
        if ser.inWaiting() != 0:
            data = ser.readline()
    except Exception as e:
        error_msg = "Error reading from {0}".format(serial_port)
        template = "An exception of type {0} occured. Arguments:\n{1!r}"
        message = template.format(type(e).__name__, e.args)

        print error_msg, message

    return data

# parse out the signal value
def parse_pending(signal_string):

    if len(signal_string) == 0:
        return

    global signal_set

    try:
    # split by ',' and get first element
        values = [int(x) for x in signal_string.split(',')]
    except:
        values = None

    # push signal into list
    if values and len(values) == 6:
        signal_set.append(values)

def moving_average_filter(last_signal, signal, size_window):
    
    signal = ([0] * ((size_window - 1) - len(last_signal))) + list(last_signal) + list(signal)
    return list(np.convolve(signal, [1.0 / size_window] * size_window, 'valid'))

def make_init_data():
    
    ret_signal = [[0] * 6] * number_of_signal
    ret_signal = zip(*ret_signal)

    ret = list()
    for label in signal_type:
        ret.append({ 'data': [p for p in enumerate([0] * number_of_signal)], 'label': label })
    ret = json.dumps({ 'signal': ret })

    return ret


def make_data():

    global signal_set
    global signal_type
    global last_signal_set
    global number_of_signal
    global size_window_MA

    # take out the signal to return
    signals = signal_set[:min(number_of_signal, len(signal_set))]
    last_signals = last_signal_set[-(min(size_window_MA - 1, len(last_signal_set))):]

    # fill the signal
    if len(signals) < number_of_signal:
        signals.extend([[0] * 6] * (number_of_signal - len(signals)))

    # transpose signals to make the signal of same type in same list
    signals = zip(*signals)
    last_signals = zip(*last_signals)

    ret = list()

    for i in xrange(6):
        if size_window_MA != 0:
            signals[i] = moving_average_filter(last_signals[i], signals[i], size_window_MA)
        ret.append({ 'data': [p for p in enumerate(signals[i])], 'label': signal_type[i] })
    ret = json.dumps({ 'signal': ret })
    return ret

# push signal data to client
def signal_tx():

    global tx_status
    global number_of_signal
    global signal_set
    global last_signal_set

    parse_pending(recieve_signal())

    if not tx_status:
        return
    
    # pop out the transmitted signal
    if len(signal_set):
        last_signal_set.append(signal_set.pop(0))
        last_signal_set = last_signal_set[-number_of_signal:]

    ret = make_data()
    for cl in client:
        cl.write_message(ret)


# tornado websocket handler
class socket_handler(tornado.websocket.WebSocketHandler):
    def open(self):
        client.append(self)
        self.write_message(make_init_data())

    def on_message(self, message):
        global tx_status
        global signal_set
        global toggle_moving_average_filter
        global size_window_MA

        token = message.split()

        if token[0] == "play":
            tx_status = True
        elif token[0] == "pause":
            tx_status = False
        elif token[0] == "clear":
            signal_set = [[0] * 6] * number_of_signal
            self.write_message(make_init_data())
        elif token[0] == "MAF":
            size_window_MA = int(token[1])
            self.write_message(make_data())

    def on_close(self):
        client.remove(self)

class homepage_handler(tornado.web.RequestHandler):
    def get(self):
        self.render('template/index.html')

settings = {
    'static_path': os.path.join(os.path.dirname(__file__), 'static'),
}

application = tornado.web.Application([
    (r'/', homepage_handler),
    (r'/ws', socket_handler),
    ], **settings)

if __name__ == "__main__":
    #tell tornado to run signal_tx every 1 ms
    serial_loop = tornado.ioloop.PeriodicCallback(signal_tx, callback_timeout)
    serial_loop.start()

    application.listen(tornado_port)
    print "Starting server on port number {0}...".format(tornado_port)
    print "Open at http://localhost:{0}/".format(tornado_port)

    try:
        tornado.ioloop.IOLoop.instance().start()
    except KeyboardInterrupt:
        print 'Server closed...'
